### DISCUSSION TITLE
Add `X.upcast` to effect-polymorphic data types

### DISCUSSION THREAD

**Discussion by mlutze** [flix/flix#8532] [Ideas]
This kind of error can be annoying:

```
mismatched effect formulas.

Type One: Iterator[Result[IOError, String], IO, r]
Type Two: Iterator[Result[IOError, String], r + IO, r]
```

There are a few ways of fixing it. We could add an upcast in the right place if we have a lambda. We could make our constructors polymorphic in their type. But we can also make the upcastability more explicit. For example, we could add `Iterator.upcast`:

```scala
def upcast(it: Iterator[a, ef, r]): Iterator[a, ef + ef1, r + r1] = ???
```

**Comment by magnus-madsen** [comment]
@JonathanStarup Will our future sub-effecting not solve this?
